lib/filewatch/watch.rb

```diff
@@ -126,77 +124,136 @@ def inode(path,stat)
       synchronized do
+        return if @files.empty?
+
+        file_deleteable = []
+        # look at the closed to see if its changed
+
+        @files.values.select {|wf| wf.closed? }.each do |watched_file|
+          path = watched_file.path
+          begin
+            stat = watched_file.restat
+            if watched_file.size_changed? || watched_file.inode_changed?(inode(path,stat))
+              # if the closed file changed, move it to the watched state
+              # not to active state because we want to use MAX_OPEN_FILES throttling.
+              watched_file.watch
+            end
+          rescue Errno::ENOENT
+            # file has gone away or we can't read it anymore.
+            file_deleteable << path
+            debug_log("each: closed: stat failed: #{path}: (#{$!}), deleting from @files")
+          rescue => e
+            debug_log("each: closed: stat failed: #{path}: (#{e.inspect})")
+          end
+        end
+
+        # look at the ignored to see if its changed
+        @files.values.select {|wf| wf.ignored? }.each do |watched_file|
+          path = watched_file.path
+          begin
+            stat = watched_file.restat
+            if watched_file.size_changed? || watched_file.inode_changed?(inode(path,stat))
+              # if the ignored file changed, move it to the watched state
+              # not to active state because we want to use MAX_OPEN_FILES throttling.
+              # this file has not been yielded to the block yet
+              # but we need to allow the tail to start from the ignored_size
+              # by adding this to the sincedb so that the subsequent modify
+              # event can detect the change
+              watched_file.watch
+              yield(:unignore, watched_file)
+            end
+          rescue Errno::ENOENT
+            # file has gone away or we can't read it anymore.
+            file_deleteable << path
+            debug_log("each: ignored: stat failed: #{path}: (#{$!}), deleting from @files")
+          rescue => e
+            debug_log("each: ignored: stat failed: #{path}: (#{e.inspect})")
+          end
+        end
+
         # Send any creates.
-        @files.each do |path, watched_file|
-          if !watched_file.create_sent?
+        if (to_take = @max_active - @files.values.count{|wf| wf.active?}) > 0
+          @files.values.select {|wf| wf.watched? }.take(to_take).each do |watched_file|
+            watched_file.activate
+            # don't do create again
+            next if watched_file.state_history_any?(:closed, :ignored)
             if watched_file.initial?
-              yield(:create_initial, path)
+              yield(:create_initial, watched_file)
             else
-              yield(:create, path)
+              yield(:create, watched_file)
             end
-            watched_file.create_sent = true
           end
         end
 
-        @files.each do |path, watched_file|
+        # warning on open file limit
```

Comment:
Let's log the number of files that we can't watch becuse it would exceed @max_active, just to let users know that this is an intentional decision by the library and not a bug.